Create 2020-11-20-SSH-Clef-Certificat (#1)
tomMoulard/blog: content/posts/2020-11-20-SSH-Certificat.md

---

> title: "Securisation SSH : Utilisation de certificat a la place des Clefs."

Reviewer: ```suggestion
title: "Sécurisation SSH : Utilisation de certificat à la place des Clefs"
```

date: 2020-11-20T13:19:36+01:00
author: Guillaume Moulard
url: /SSH-Certificat
draft: false
type: post
tags:
  - SSH
  - Securite
  - certificat
  - sshd
categories:
  - tutoriel
---

> # Pourquoi les certificats SSL C'est mieux qu'une clefs SSH

Reviewer: ```suggestion
# Pourquoi les certificats SSL c'est mieux qu'une clefs SSH ?
```

> Mettre en place des certificats pousse encore la securité un peut plus loins car il est alors possible de donnée des delais de validité de faire les liste de revocation et des restriction d'usage du SSH : clear,force-command=command, no-agent-forwarding, no-port-forwarding, no-pt, no-user-rc, no-x11-forwarding, permit-agent-forwarding, permit-port-forwarding, permit-pty, permit-user-rc, permit-x11-forwarding, source-address=address_list

Reviewer: ```suggestion
Mettre en place des certificats pousse encore la sécurité un peu plus loin car il est alors possible:

 - de donner des délais de validité
 - de faire les liste de révocation
 - de faire des restriction d'usage du SSH (`clear`, `force-command=command`, `no-agent-forwarding`, `no-port-forwarding`, `no-pt`, `no-user-rc`, `no-x11-forwarding`, `permit-agent-forwarding`, `permit-port-forwarding`, `permit-pty`, `permit-user-rc`, `permit-x11-forwarding`, `source-address=address_list`)
```

> # Comment faire mettre en oeuvre une CA sur le serveur SSHD 

Reviewer: ```suggestion
# Comment mettre en oeuvre une CA sur le serveur SSHD 
```

> ## 1. creation de la CA Certificat Autority

Reviewer: ```suggestion
## 1. Création de la CA (Certificat Autority)
```

```bash
root@pi3:~# ssh-keygen -t rsa -b 4096 -f /etc/ssh/ca
Generating public/private rsa key pair.
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in /etc/ssh/ca.
Your public key has been saved in /etc/ssh/ca.pub.
The key fingerprint is:
SHA256:Ax5NHQgoY8kpb3lNA/zElBFvJhQAt1G+EsJfMXyHPdo root@pi3
The key's randomart image is:
+---[RSA 4096]----+
| ..==O@*.=..     |
|..B.o=B*+ =      |
| +o+o=*o=+ .     |
|  +o.+oB. E      |
| . .o o S        |
|     .   .       |
|                 |
|                 |
|                 |
+----[SHA256]-----+
root@pi3:~# ls -l /etc/ssh/ca*
-rw------- 1 root root 3243 mars   7 12:33 /etc/ssh/ca
-rw-r--r-- 1 root root  734 mars   7 12:33 /etc/ssh/ca.pub

> root@pi3:~#

Reviewer: ```suggestion
```

```bash

> ## 2. setup des la CA dans les serveur SSH sshd

Reviewer: ```suggestion
## 2. Utilisation des CA dans les serveur SSH sshd
```

```bash
echo "TrustedUserCAKeys /etc/ssh/ca.pub" >> /etc/ssh/sshd_config
root@pi3:~# systemctl restart sshd
``` 

> # Comment créer sa clefs et la signé par sa CA

Reviewer: ```suggestion
# Comment créer sa clefs et la signer avec le CA
```

> comme precedement, pour créer sa clef, rien de plus simple: 

Reviewer: ```suggestion
Comme précédemment, pour créer sa clef, rien de plus simple:
```

> avec openssh c'est encore plus simple 

Reviewer: Il en manque un bout non ?

```suggestion
Avec openssh c'est très simple 
```

```bash

> pi@pi3:~/.ssh $ ssh-keygen -t rsa -b 4096 -f /home/pi/.ssh/clef2

Reviewer: ```suggestion
pi@pi3:~/.ssh $ ssh-keygen -t rsa -b 4096 -f /home/pi/.ssh/cle2
```

Generating public/private rsa key pair.
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in /home/pi/.ssh/clef2.

> Your public key has been saved in /home/pi/.ssh/clef2.pub.

Reviewer: ```suggestion
Your identification has been saved in /home/pi/.ssh/cle2.
Your public key has been saved in /home/pi/.ssh/cle2.pub.
```

The key fingerprint is:
SHA256:AAcbm67DjpfcQOS9Ii8JS3H1ZPs9MsCmKA/jXACEeF0 pi@pi3
The key's randomart image is:
+---[RSA 4096]----+
|+. .+oE          |
|+ o .O o         |
|.+ .= * .        |
| ooo.  B         |
| .+ o.o S .      |
|o*o+..   + o     |
|=*X=      o .    |
|+==o.            |
|.o.              |
+----[SHA256]-----+

> pi@pi3:~/.ssh $

Reviewer: ```suggestion
```

```

et la signature par le CA :

```bash

> pi@pi3:~/.ssh $ sudo ssh-keygen -s /etc/ssh/ca -n "Signaturede la clef" -I 1 -V +365d /home/pi/.ssh/clef2.pub

Reviewer: ```suggestion
pi@pi3:~/.ssh $ sudo ssh-keygen -s /etc/ssh/ca -n "Signature de la clef" -I 1 -V +365d /home/pi/.ssh/clef2.pub
```

> Signed user key /home/pi/.ssh/clef2-cert.pub: id "1" serial 0 for Signaturede la clef valid from 2020-03-07T12:49:00 to 2021-03-07T12:50:00

Reviewer: ```suggestion
Signed user key /home/pi/.ssh/clef2-cert.pub: id "1" serial 0 for Signature de la clef valid from 2020-03-07T12:49:00 to 2021-03-07T12:50:00
```

> pi@pi3:~/.ssh $ sudo chown pi:pi /home/pi/.ssh/clef2-cert.pub

Reviewer: ```suggestion
pi@pi3:~/.ssh $ sudo chown pi:pi /home/pi/.ssh/cle2-cert.pub
```

```

> Ont a mainenant : un ficher /home/pi/.ssh/clef2-cert.pub qui est comme une clef public classique mais avec une duré de validité de 365 jour

Reviewer: ```suggestion
On a maintenant un ficher `/home/pi/.ssh/clef2-cert.pub` qui est comme une clef publique classique mais avec une durée de validité de 365 jours.
```

```bash

> pi@pi3:~/.ssh $ cat /home/pi/.ssh/clef2-cert.pub

Reviewer: ```suggestion
pi@pi3:~/.ssh $ cat /home/pi/.ssh/cle2-cert.pub
```

```

> # Comment pousser sa clef signé sur un serveur

Reviewer: ```suggestion
# Comment pousser sa clef signée sur un serveur
```

> Exactement comme pour une clefs public, il faut la mettre dans le fichier autorizekey

Reviewer: ```suggestion
Exactement comme pour une clefs publique, il faut la mettre dans le fichier `~/.ssh/authorized_keys`.
```

```bash
pi@pi3:~/.ssh $ ssh-copy-id -f -i /home/pi/.ssh/clef2-cert.pub pi@pi4b.home

> /usr/bin/ssh-copy-id: INFO: Source of key(s) to be installed: "/home/pi/.ssh/clef2-cert.pub"

Reviewer: ```suggestion
pi@pi3:~/.ssh $ ssh-copy-id -f -i /home/pi/.ssh/cle2-cert.pub pi@pi4b.home
/usr/bin/ssh-copy-id: INFO: Source of key(s) to be installed: "/home/pi/.ssh/cle2-cert.pub"
```

no such identity: /home/pi/.ssh/identity: No such file or directory

Number of key(s) added: 1

Now try logging into the machine, with:   "ssh 'pi@pi4b.home'"
and check to make sure that only the key(s) you wanted were added.

> pi@pi3:~/.ssh $ 

Reviewer: ```suggestion
```

```

```bash

> pi@pi3:~/.ssh $ ssh -i clef2  pi@pi4b.home

Reviewer: ```suggestion
pi@pi3:~/.ssh $ ssh -i cle2  pi@pi4b.home
```

no such identity: /home/pi/.ssh/identity: No such file or directory
Linux pi4b 4.19.75-v7l+ #1270 SMP Tue Sep 24 18:51:41 BST 2019 armv7l

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.

> Last login: Sat Mar  7 11:57:23 2020 from 192.168.1.29

Reviewer: Ca ne vaudrait pas le coup d'enlever ton @ip ?
